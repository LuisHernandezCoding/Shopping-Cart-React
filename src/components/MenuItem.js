import { PropTypes } from 'prop-types';
import { useState } from 'react';
import './MenuItem.css';

const MenuItem = ({ item, addToCart }) => {
  const [quantity, setQuantity] = useState(1);

  const {
    id,
    name,
    price,
    image,
    ingredients,
  } = item;

  return (
    <div
      key={id}
      className="column is-quarter animate__animated animate__delay-1s animate__zoomIn"
    >
      <div className="card">
        <div className="card-image">
          <figure className="image is-4by3">
            <img src={image} alt={name} />
          </figure>
        </div>
        <div className="card-content">
          <div className="media">
            <div className="media-content">
              <p className="title is-4">{name}</p>
              <p className="subtitle is-6">
                <span className="icon">
                  <i className="fas fa-dollar-sign" />
                </span>
                {price}
              </p>
            </div>
          </div>
          <div className="content">
            {ingredients}
            <hr className="menu-item-divider has-background-grey-lighter" />
            <div className="buttons is-right">
              <input type="number" className="column is-half input button is-small" min="1" max="10" defaultValue="1" onChange={(e) => setQuantity(e.target.value)} />
              <button type="button" className="button is-success" onClick={() => addToCart(item, quantity)}>
                Add to cart
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

MenuItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    ingredients: PropTypes.string.isRequired,
  }).isRequired,
  addToCart: PropTypes.func.isRequired,
};

export default MenuItem;
