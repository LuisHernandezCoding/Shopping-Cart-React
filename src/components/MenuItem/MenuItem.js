import { PropTypes } from 'prop-types';
import { useState } from 'react';

const MenuItem = ({ item, itemMethods, cart }) => {
  const [quantity, setQuantity] = useState(1);

  const {
    addItemToCart,
    removeItemFromCart,
    updateItemQuantity,
  } = itemMethods;

  const {
    id,
    name,
    price,
    image,
    ingredients,
  } = item;

  return (
    <div
      className="animate__animated animate__delay-1s animate__zoomIn"
    >
      <div className="card item has-background-black has-text-white">
        <div className="card-image">
          <figure className="image is-4by3">
            <img src={image} alt={name} />
          </figure>
        </div>
        <div className="card-content">
          <div className="media">
            <div className="media-content">
              <p className="title is-5 has-text-white is-flex is-justify-content-space-between">
                {name}
                <span className="has-text-white is-flex is-align-items-center">
                  <span className="icon">
                    <i className="fas fa-dollar-sign" />
                  </span>
                  {price}
                </span>
              </p>
            </div>
          </div>
          <div className="content">
            {ingredients}
            <hr className="menu-item-divider has-background-grey-lighter" />
            {cart.find((item) => item.id === id) ? (
              <div className="field has-addons">
                <div className="control">
                  <button
                    type="button"
                    className="button is-danger"
                    onClick={() => {
                      if (cart.find((item) => item.id === id).quantity > 1) {
                        // make async call to update quantity
                        const newQuantity = cart.find((item) => item.id === id).quantity - 1;
                        setQuantity(newQuantity);
                        updateItemQuantity(item, newQuantity);
                      } else {
                        removeItemFromCart(item);
                      }
                    }}
                  >
                    -
                  </button>
                </div>
                <div className="control">
                  <input
                    className="input"
                    type="text"
                    value={cart.find((item) => item.id === id).quantity}
                    readOnly
                  />
                </div>
                <div className="control">
                  <button
                    type="button"
                    className="button is-success"
                    onClick={
                      () => {
                        setQuantity(cart.find((item) => item.id === id).quantity + 1);
                        updateItemQuantity(item, cart.find((item) => item.id === id).quantity + 1);
                      }
                    }
                  >
                    +
                  </button>
                </div>
              </div>
            ) : (
              <div className="buttons is-right">
                <input
                  type="number"
                  className="column is-half input button is-small"
                  min="1"
                  max="10"
                  defaultValue="1"
                  onChange={(e) => setQuantity(e.target.value)}
                />
                <button
                  type="button"
                  className="button is-success"
                  onClick={() => addItemToCart({
                    id,
                    name,
                    price,
                    quantity,
                    image,
                    ingredients,
                  })}
                >
                  Add to cart
                </button>
              </div>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

MenuItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    ingredients: PropTypes.string.isRequired,
  }).isRequired,
  itemMethods: PropTypes.shape({
    addItemToCart: PropTypes.func.isRequired,
    removeItemFromCart: PropTypes.func.isRequired,
    updateItemQuantity: PropTypes.func.isRequired,
  }).isRequired,
  cart: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    quantity: PropTypes.number.isRequired,
  })).isRequired,
};

export default MenuItem;
