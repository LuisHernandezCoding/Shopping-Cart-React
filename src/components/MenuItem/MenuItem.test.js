import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import MenuItem from './MenuItem';

afterEach(cleanup);

const item = {
  id: 1,
  name: 'Test Item',
  price: 10,
  image: 'test.jpg',
  ingredients: 'Test ingredients',
  quantity: 1,
};

const cart = [];

const itemMethods = {
  addItemToCart: jest.fn(),
  removeItemFromCart: jest.fn(),
  updateItemQuantity: jest.fn(),
};

describe('MenuItem', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <MenuItem item={item} itemMethods={itemMethods} cart={cart} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('addItemToCart should be called when Add to cart button is clicked', () => {
    const { getByText } = render(
      <MenuItem item={item} itemMethods={itemMethods} cart={cart} />,
    );
    fireEvent.click(getByText('Add to cart'));
    expect(itemMethods.addItemToCart).toHaveBeenCalled();
  });

  it('addItemToCart should be called with the correct quantity when Add to cart button is clicked', () => {
    const { getByText } = render(
      <MenuItem item={item} itemMethods={itemMethods} cart={cart} />,
    );
    const addCartButton = getByText('Add to cart');
    const input = addCartButton.previousSibling;
    fireEvent.change(input, { target: { value: 2 } });
    fireEvent.click(addCartButton);
    expect(itemMethods.addItemToCart).toHaveBeenCalledWith({
      id: 1,
      name: 'Test Item',
      price: 10,
      image: 'test.jpg',
      ingredients: 'Test ingredients',
      quantity: '2',
    });
  });

  it('removeItemFromCart should be called when - button is clicked', () => {
    cart.push(item);
    const { getByText } = render(
      <MenuItem item={item} itemMethods={itemMethods} cart={cart} />,
    );
    fireEvent.click(getByText('-'));
    expect(itemMethods.removeItemFromCart).toHaveBeenCalled();
  });

  it('updateItemQuantity should be called when + button is clicked', () => {
    cart.push(item);
    const { getByText } = render(
      <MenuItem item={item} itemMethods={itemMethods} cart={cart} />,
    );
    fireEvent.click(getByText('+'));
    expect(itemMethods.updateItemQuantity).toHaveBeenCalled();
  });
});
