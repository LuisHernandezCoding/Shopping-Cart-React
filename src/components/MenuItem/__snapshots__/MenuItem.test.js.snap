// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem should render correctly 1`] = `
<DocumentFragment>
  <div
    class="animate__animated animate__delay-1s animate__zoomIn"
  >
    <div
      class="card item has-background-black has-text-white"
    >
      <div
        class="card-image"
      >
        <figure
          class="image is-4by3"
        >
          <img
            alt="Test Item"
            src="test.jpg"
          />
        </figure>
      </div>
      <div
        class="card-content"
      >
        <div
          class="media"
        >
          <div
            class="media-content"
          >
            <p
              class="title is-5 has-text-white is-flex is-justify-content-space-between"
            >
              Test Item
              <span
                class="has-text-white is-flex is-align-items-center"
              >
                <span
                  class="icon"
                >
                  <i
                    class="fas fa-dollar-sign"
                  />
                </span>
                10
              </span>
            </p>
          </div>
        </div>
        <div
          class="content"
        >
          Test ingredients
          <hr
            class="menu-item-divider has-background-grey-lighter"
          />
          <div
            class="buttons is-right"
          >
            <input
              class="column is-half input button is-small"
              max="10"
              min="1"
              type="number"
              value="1"
            />
            <button
              class="button is-success"
              type="button"
            >
              Add to cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
